src/dhcpmon/src/dhcp_device.cpp

```diff
@@ -163,86 +213,92 @@ static void handle_dhcp_option_53(dhcp_device_context_t *context,
+{
+    struct ether_header *ethhdr = (struct ether_header*) context->buffer;
+    struct ip *iphdr = (struct ip*) (context->buffer + IP_START_OFFSET);
+    struct udphdr *udp = (struct udphdr*) (context->buffer + UDP_START_OFFSET);
+    uint8_t *dhcphdr = context->buffer + DHCP_START_OFFSET;
+    int dhcp_option_offset = DHCP_START_OFFSET + DHCP_OPTIONS_HEADER_SIZE;
+
+    if (((unsigned)buffer_sz > UDP_START_OFFSET + sizeof(struct udphdr) + DHCP_OPTIONS_HEADER_SIZE) &&
+        (ntohs(udp->len) > DHCP_OPTIONS_HEADER_SIZE))
+    {
+        int dhcp_sz = ntohs(udp->len) < buffer_sz - UDP_START_OFFSET - sizeof(struct udphdr) ?
+                    ntohs(udp->len) : buffer_sz - UDP_START_OFFSET - sizeof(struct udphdr);
+        int dhcp_option_sz = dhcp_sz - DHCP_OPTIONS_HEADER_SIZE;
+        const u_char *dhcp_option = context->buffer + dhcp_option_offset;
+        dhcp_packet_direction_t dir = (ethhdr->ether_shost[0] == context->mac[0] &&
+                                    ethhdr->ether_shost[1] == context->mac[1] &&
+                                    ethhdr->ether_shost[2] == context->mac[2] &&
+                                    ethhdr->ether_shost[3] == context->mac[3] &&
+                                    ethhdr->ether_shost[4] == context->mac[4] &&
+                                    ethhdr->ether_shost[5] == context->mac[5]) ?
+                                    DHCP_TX : DHCP_RX;
+        int offset = 0;
+        while ((offset < (dhcp_option_sz + 1)) && dhcp_option[offset] != 255) {
+            if (dhcp_option[offset] == OPTION_DHCP_MESSAGE_TYPE) {
+                if (offset < (dhcp_option_sz + 2)) {
+                    handle_dhcp_option_53(context, &dhcp_option[offset], dir, iphdr, dhcphdr);
+                }
+                break; // break while loop since we are only interested in Option 53
+            }
+
+            if (dhcp_option[offset] == 0) { // DHCP Option Padding
+                offset++;
+            } else {
+                offset += dhcp_option[offset + 1] + 2;
+            }
+        }
+    } else {
+        syslog(LOG_WARNING, "read_callback(%s): read length (%ld) is too small to capture DHCP options",
+            context->intf, buffer_sz);
+    }
+}
+
+/**
+ * @code read_tx_callback(fd, event, arg);
  *
- * @brief callback for libevent which is called every time out in order to read queued packet capture
+ * @brief callback for libevent which is called every time out in order to read queued outgoing packet capture
  *
  * @param fd            socket to read from
  * @param event         libevent triggered event
  * @param arg           user provided argument for callback (interface context)
  *
  * @return none
  */
-static void read_callback(int fd, short event, void *arg)
+static void read_tx_callback(int fd, short event, void *arg)
 {
     dhcp_device_context_t *context = (dhcp_device_context_t*) arg;
     ssize_t buffer_sz;
 
     while ((event == EV_READ) &&
```

Comment:
Here "event == EV_READ" will always true as we register this handler for EV_READ, we don't need to check it again.

Reply:
event could be EV_PERIST or EV_READ

Reply:
EV_PERIST controls the event is persistent, it makes you don't need to use event_add() again after callback. But when callback is called it must be a READ event, as you don't register read_tx_callback to any WRITE/DELETE events.

Reply:
Thanks I see. Removed event check